lib/bridge/MemberListSyncer.js

```diff
@@ -247,10 +267,86 @@ function joinMatrixUsersToChannels(rooms, server, injectJoinFn) {
     return d.promise;
 }
 
-function leaveIrcUsersFromRooms(rooms, server) {
-    return Promise.resolve();
+MemberListSyncer.prototype.leaveIrcUsersFromRooms = function(rooms, server) {
+    log.info(
+        `leaveIrcUsersFromRooms: storing member list info for ${rooms.length} ` +
+        `rooms for server ${server.domain}`
+    );
+
+    // Store the matrix room info in memory for later retrieval when NAMES is received
+    // and updateIrcMemberList is called. At that point, we have enough information to
+    // leave users from the channel that the NAMES is for.
+    rooms.forEach((roomInfo) => {
+        this._memberLists.matrix[roomInfo.id] = roomInfo;
+    });
 }
 
+// Update the MemberListSyncer with the IRC NAMES_RPL that has been received for channel.
+// This will leave any matrix users that do not have their associated IRC nick in the list
+// of names for this channel.
+MemberListSyncer.prototype.updateIrcMemberList = Promise.coroutine(function*(channel, names) {
+    log.info(`updateIrcMemberList: Updating IRC member list for ${channel}`);
+    if (this._memberLists.irc[channel] !== undefined) {
+        log.info(`Names already received for ${channel}`);
+        return;
+    }
+    if (!this.server.shouldSyncMembershipToMatrix("initial", channel)) {
+        log.info(`Initial membership list syncing disabled for ${channel} and/or globally`);
+        return;
+    }
+
+    this._memberLists.irc[channel] = Object.keys(names);
+
```

Comment:
I would log something on this line to say how many IRC users you're looking at, what channel/server, etc *before* you do all the work so we can gauge how long it takes and can correlate logs if something below goes wrong.